### PR TITLE
 SNI based Configuration for traffic server

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

**Comment by persiaAziz** [comment]
[approve ci autest]

**Comment by persiaAziz** [comment]
[approve ci autest]

File: iocore/net/ActionProcessor.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+void
+ActionProcessor::PerformAction(Continuation *cont, cchar *servername)
+{
+  SNIConfig::scoped_config params;
+  auto actionvec = params->get(servername);
+  if (!actionvec)
+    Debug("ssl_sni", "%s not available in the map", servername);
+  else {
+    for (auto it : *actionvec) {
+      ActionItem *ai = it;
```
Comment (SolidWallOfCode) [inline iocore/net/ActionProcessor.cc]: What's the purpose of `ai` here, instead of just using `it`?

File: iocore/net/P_SSLSNI.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  NextHopProperty();
+};
+
+typedef std::vector<ActionItem *> actionVector;
+typedef HashMap<cchar *, StringHashFns, actionVector *> SNIMap;
+typedef HashMap<cchar *, StringHashFns, NextHopProperty *> NextHopPropertyTable;
+
+struct SNIConfigParams : public ConfigInfo {
+  char *sni_filename = nullptr;
+  mutable SNIMap sni_action_map;
```
Comment (SolidWallOfCode) [inline iocore/net/P_SSLSNI.h]: I'm still concerned about these being `mutable`. Shouldn't all of these be set up during configuration and not changed while the configuration is live?

File: iocore/net/P_SSLUtils.h [hunk 3]
```diff
@@ -215,6 +217,25 @@ namespace detail
 /* namespace ssl */ } /* namespace detail */
 }
 
+struct ats_wildcard_matcher {
```
Comment (SolidWallOfCode) [inline iocore/net/P_SSLUtils.h]: Why were these moved up to the header from the source file? To make them inlined?

File: iocore/net/ActionProcessor.cc [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#include "ts/ink_memory.h"
+#include "P_SSLSNI.h"
+#include "P_Net.h"
+#include "P_SSLNextProtocolAccept.h"
+#include "P_SSLUtils.h"
+
+extern Map<int, SSLNextProtocolSet *> snpsMap;
+
+void
+ActionProcessor::PerformAction(Continuation *cont, cchar *servername)
```
Comment (SolidWallOfCode) [inline iocore/net/ActionProcessor.cc]: Unfortunately this name is likely to get bonged, because of class `Action` in `iocore/eventsystem/I_Action.h`.

File: iocore/net/SSLSNIConfig.cc [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+  if (0 == strncasecmp(actionTag, "TS_VERIFY_SERVER", sizeof("TS_VERIFY_SERVER")))
+    return TS_VERIFY_SERVER;
+  if (0 == strncasecmp(actionTag, "TS_CLIENT_CERT", sizeof("TS_CLIENT_CERT")))
+    return TS_CLIENT_CERT;
+  Warning("%s is not a valid action", actionTag);
+  return -1;
+}
+
+void
+SNIConfigParams::setPropertyConfig(int id, char *servername, void *param)
```
Comment (SolidWallOfCode) [inline iocore/net/SSLSNIConfig.cc]: Should `id` be `AllActions` instead of `int`?

File: lib/ts/Vec.h [hunk 6]
```diff
@@ -853,9 +853,9 @@ template <class C, class A, int S>
 inline void
 Vec<C, A, S>::free_and_clear()
 {
-  for (int x = 0; x < n; x++)
+  for (int x = 0; x < static_cast<int>(n); x++)
     if (v[x])
```
Comment (SolidWallOfCode) [inline lib/ts/Vec.h]: Why check `v[x]`? It's legal to pass `nullptr` to `free`.

File: lib/ts/Vec.h [hunk 7]
```diff
@@ -853,9 +853,9 @@ template <class C, class A, int S>
 inline void
 Vec<C, A, S>::free_and_clear()
 {
-  for (int x = 0; x < n; x++)
+  for (int x = 0; x < static_cast<int>(n); x++)
```
Comment (SolidWallOfCode) [inline lib/ts/Vec.h]: Why not declare `x` of a type that doesn't require `n` to be cast?

File: lib/ts/Vec.h [hunk 8]
```diff
@@ -853,9 +853,9 @@ template <class C, class A, int S>
...
     if (v[x])
-      A::free(v[x]);
+      A::free((void *)v[x]);
```
Comment (SolidWallOfCode) [inline lib/ts/Vec.h]: Why cast? If it's to remove `const` this is precisely why C++ style casts are strongly encouraged - your intent would then be clear rather than ambiguous.

**Comment by SolidWallOfCode** [comment]
Please clear up the merge conflicts. Also, we will probably want some documentation beyond the comments in the default configuration file.

File: iocore/net/P_SSLSNI.h (hunk 2 above)
Comment (persiaAziz) [inline iocore/net/P_SSLSNI.h]: The reason this is mutable is because all the getter functions of HashMap or Map in lib/ts are non-const so I can't pass the keys (which are cchar*) to obtain the mapped objects.

**Comment by persiaAziz** [comment]
[approve ci autest]

File: lib/ts/Vec.h (hunk 6 above)
Comment (persiaAziz) [inline lib/ts/Vec.h]: because of this : 
../../lib/ts/Vec.h:858:16: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]

File: lib/ts/Vec.h (hunk 8 above)
Comment (persiaAziz) [inline lib/ts/Vec.h]: ../../lib/ts/Vec.h:858:16: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]

File: iocore/net/P_SSLUtils.h (hunk 3 above)
Comment (persiaAziz) [inline iocore/net/P_SSLUtils.h]: yah and also this should have been in the header file so I could access these from SSLSNIConfig

File: cmd/traffic_layout/traffic_layout.cc [hunk 9]
```diff
@@ -167,6 +167,7 @@ produce_layout(bool json)
   print_var("remap.config", RecConfigReadConfigPath("proxy.config.url_remap.filename"), json);
   print_var("plugin.config", RecConfigReadConfigPath(nullptr, "plugin.config"), json);
   print_var("ssl_multicert.config", RecConfigReadConfigPath("proxy.config.ssl.server.multicert.filename"), json);
+  print_var("ssl_SNI.config", RecConfigReadConfigPath("proxy.config.ssl.SNI.filename"), json);
```
Comment (zwoop) [inline cmd/traffic_layout/traffic_layout.cc]: Not hugely favorable to the upper case here, but there are a handful of existing ones (.CA.) already...

File: cmd/traffic_manager/AddConfigFilesHere.cc [hunk 10]
```diff
@@ -77,5 +77,6 @@ initializeRegistry()
   configFiles->addFile("splitdns.config", false);
   configFiles->addFile("ssl_multicert.config", false);
   configFiles->addFile("metrics.config", false);
+  configFiles->addFile("ssl_SNI.config", false);
```
Comment (zwoop) [inline cmd/traffic_manager/AddConfigFilesHere.cc]: I feel pretty strongly that we should *not* have upper case letters in the filename however.

Also, this being a new configuration file, I wonder if we should mandate that any new configuration is using the Lua format? @SolidWallOfCode  thoughts?

I'd recommend just **sni.lua**, **sslsni.lua**, **ssl_sni.lua** or perhaps s/.lua/.luaconfig/ if that's where we go. Such a change though (to go to Lua) would require significant rework here, but I think it's worth it to set the stage clear for new configurations.

File: iocore/net/ActionProcessor.cc [hunk 11]
```diff
@@ -0,0 +1,53 @@
+/** @file
```
Comment (zwoop) [inline iocore/net/ActionProcessor.cc]: Hmmm, why is this a Processor? Processors have pretty significant semantics in our code, such that a Processor is heavily involved with the event processing. This is not that, is it ?

File: cmd/traffic_manager/AddConfigFilesHere.cc (hunk 10 above)
Comment (persiaAziz) [inline cmd/traffic_manager/AddConfigFilesHere.cc]: Is there any Lua example I can look at?

File: iocore/net/ActionProcessor.cc (hunk 11 above)
Comment (persiaAziz) [inline iocore/net/ActionProcessor.cc]: well, it processes the SNI based actions. I am changing the class name anyway

File: cmd/traffic_manager/AddConfigFilesHere.cc (hunk 10 above)
Comment (zwoop) [inline cmd/traffic_manager/AddConfigFilesHere.cc]: Yes, metric.config and logging.config were both migrated to Lua.

**Comment by bryancall** [comment]
@persiaAziz
Please reopen with a Lua configuration file format.
